### PR TITLE
Issue #8216 - make processing of metadata in OpenIdConfiguration extensible

### PR DESCRIPTION
**Pull request by lachlan-roberts**
## Issue #8216

Make `processMetadata()` and `fetchOpenIdConnectMetadata()` protected inside `OpenIdConfiguration` to make it easier to extend.

### CODE REVIEW

**Review by joakime — DISMISSED** [review]

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java [hunk 1]
```diff
@@ -116,35 +116,33 @@ protected void doStart() throws Exception
...
         }
     }
 
-    private static HttpClient newHttpClient()
+    protected void processMetadata(Map<String, Object> discoveryDocument)
     {
-        ClientConnector connector = new ClientConnector();
-        connector.setSslContextFactory(new SslContextFactory.Client(false));
-        return new HttpClient(new HttpClientTransportOverHTTP(connector));
+        authEndpoint = (String)discoveryDocument.get("authorization_endpoint");
```
Comment (gregw) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java]: Make "authorization_endpoint" and "token_endpoint" be constant Strings

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java [hunk 2]
```diff
@@ -116,35 +116,33 @@ protected void doStart() throws Exception
...
-
-            if (!Objects.equals(discoveryDocument.get("issuer"), issuer))
-                LOG.warn("The issuer in the metadata is not correct.");
+            Map<String, Object> discoveryDocument = fetchOpenIdConnectMetadata();
+            processMetadata(discoveryDocument);
         }
     }
 
-    private static HttpClient newHttpClient()
+    protected void processMetadata(Map<String, Object> discoveryDocument)
```
Comment (gregw) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java:133]: javadoc

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java [hunk 3]
```diff
@@ -116,35 +116,33 @@ protected void doStart() throws Exception
...
+        tokenEndpoint = (String)discoveryDocument.get("token_endpoint");
+        if (tokenEndpoint == null)
+            throw new IllegalArgumentException("token_endpoint");
+
+        if (!Objects.equals(discoveryDocument.get("issuer"), issuer))
+            LOG.warn("The issuer in the metadata is not correct.");
     }
 
-    private static Map<String, Object> fetchOpenIdConnectMetadata(String provider, HttpClient httpClient)
+    protected Map<String, Object> fetchOpenIdConnectMetadata()
```
Comment (gregw) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java:153]: javadoc

**Review by gregw — CHANGES_REQUESTED** [review]

File: jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java [hunk 4]
```diff
@@ -121,20 +124,27 @@ protected void doStart() throws Exception
         }
     }
 
+    /**
+     * Process the OpenID Connect metadata discovered by {@link #fetchOpenIdConnectMetadata()}.
```
Comment (gregw) [inline jetty-openid/src/main/java/org/eclipse/jetty/security/openid/OpenIdConfiguration.java:128]: I was kind of hoping that you'd explain what processing can be done and what is done by default.
Currently it looks like you are just validating some parameters exist or are a certain value, but not much else.

So you need to at leaset explain that IAE can be  thrown, perhaps even add it to the signature

**Review by gregw — CHANGES_REQUESTED** [review]

**Review by gregw — APPROVED** [review]

**Pull request merged**